ai_eval/coach.py

```diff
@@ -0,0 +1,817 @@
+""").strip()  # noqa
+
+
+DEFAULT_EVALUATOR_PROMPT = textwrap.dedent("""
+    You are an evaluator agent responsible for generating an evaluation report of the conversation after the conversation has concluded.
+    Use the provided chat history to evaluate the learner based on the evaluation criteria.
+    You are evaluating the user based on their input, not the reactions by the other characters (such as the main character or the coach).
+    **Important**: Your only job is to give an evaluation report in well-structured markdown. You are not to chat with the learner. Do not engage in any conversation or provide feedback directly to the user. Do not ask questions, give advice or encouragement, or continue the conversation. Your only job is to produce the evaluation report.
+    Your task is to produce a well-structured markdown report in the following format:
+
+    # Evaluation Report
+
+    {% for criterion in scenario_data.evaluation_criteria %}
+        ## {{ criterion.name }}
+        ### Score: (0-5)/5
+        **Rationale**: Provide a rationale for the score, using specific direct quotes from the conversation as evidence.
+    {% endfor %}
+
+    Your response must adhere to this exact structure, and each score must have a detailed rationale that includes at least one direct quote from the chat history.
+    If you cannot find a direct quote, mention this explicitly and provide an explanation.
+""").strip()  # noqa
+
+
+DEFAULT_CONVERSATION_FORMAT = textwrap.dedent("""
+    <conversation>
+        {% for message in messages %}
+            <message>
+                <agent>{{ message.character.name }}</agent>
+                <role>{{ message.character.role }}</role>
+                <content>{{ message.content | escape }}</content>
+            </message>
+        {% endfor %}
+    </conversation>
+""")
+
+
+class CoachAIEvalXBlock(AIEvalXBlock):
+    """
+
+    AI-powered XBlock for simulated conversations with
+    two simulated characters.
+
+    """
+
+    _jinja_env = SandboxedEnvironment(undefined=jinja2.StrictUndefined)
+
+    display_name = String(
+        display_name=_("Display Name"),
+        help=_("Name of the component in the studio"),
+        default="Coached AI Evaluation",
+        scope=Scope.settings,
+    )
+
+    evaluator_prompt = String(
+        display_name=_("Evaluator prompt"),
+        help=_(
+            "Prompt used to instruct the model how to evaluate the learner"
+        ),
+        multiline_editor=True,
+        default=DEFAULT_EVALUATOR_PROMPT,
+        scope=Scope.settings,
+    )
+
+    scenario_title = String(
+        display_name=_("Scenario title"),
+        default="",
+        scope=Scope.settings,
+    )
+
+    initial_message = String(
+        display_name=_("Initial message"),
+        default="",
+        scope=Scope.settings,
+    )
+
+    scenario_data = Dict(
+        display_name=_("Scenario data"),
+        help=_("Arbitrary data accessible in prompt templates"),
```

Comment:
This needs more docs:
- what's the data format?
- examples of accessing in prompt templates?
- which fields can access these?

Also, is this necessary? What's stopping instructors from inlining these? What advantage do we get by using templating here? This feels very technical for instructors - they need to know JSON, Jinja templating, and they don't have an easy way to debug any issues in the templates. A broken template will be accepted and result in a blind 500 error on submission.